### ISSUE TITLE
Nuster + EFS

### ISSUE THREAD

**Issue by igorescobar** [jiangwenyuan/nuster#76]
Hi,

I'm sorry if this is not the right place to ask this question. AWS has a service called EFS (Amazon Elastic File System). With it, you can mount the same volume into several docker containers for example. It means that, in theory, I could have several application containers, one or more Nuster instances sharing the same volume for caching.

Using ECS + EFS I could even have one dedicated container for Nuster acting as a reverse proxy and another one just for the application. It means that during an auto-scaling I could have many Nuster containers pointing to the same volume during a fraction of time.

How would Nuster perform in this scenario, would that be a problem?

Thanks!

**Comment by jiangwenyuan** [comment]
Hi, after a quick though, I felt it should probably work, since I don't have EFS environment, do you mind to have some test?

**Comment by packeteer** [comment]
faire, EFS is essentially NFS

**Comment by igorescobar** [comment]
I'm asking because like Nginx for example, it builds a in-memory index with all the cached files inside your cache folder and it uses that in-memory table to decide if it is a MISS or a HIT. 

So during the initialisation of the container (or several containers at the same time) it results into a ton of MISSES and sometimes it even gets lost and starts removing the entire cache folder and starts rebuilding it again.

So in order to minimise this "destroy cache folder" effect we literally had to decouple the nginx from the api container network and deploy is as a standalone container with its own lifecyle process.

So my question is, Does Nuster have the same in-memory table? if it does how does Nuster builds it and how would it perform on a situation described above where you have several in-memory different states pointing to the same volume.

**Comment by igorescobar** [comment]
I'm asking because like Nginx for example, it builds a in-memory index with all the cached files inside your cache folder and it uses that in-memory table to decide if it is a MISS or a HIT. 

So during the initialisation of the container (or several containers at the same time) it results into a ton of MISSES and sometimes it even gets lost and starts removing the entire cache folder and starts rebuilding it again.

So in order to minimise this "destroy cache folder" effect we literally had to decouple the nginx from the api container network and deploy is as a standalone container with its own lifecyle process.

So my question is, Does Nuster have the same in-memory table? if it does how does Nuster builds it and how would it perform on a situation described above where you have several in-memory different states pointing to the same volume.

**Comment by jiangwenyuan** [comment]
@igorescobar I see, nuster does have an in-memory table but act in a different way.

There are two stores: memory and disk, and four modes:

- memory on disk on
- memory on disk  off
- memory on disk sync
- memory off disk on

And act differently if nuster restarts and loaded or not.

1. not restart

For `memory on disk on|off|sync`, nuster just use the cache stored in memory to serve clients.

For `memory off disk on`, nuster use the in-memory table to get the cache file and serve clients.

2. restart

All data in in-memory table lost, so after restart, there's a load process to load necessary info to rebuild the in-memory table: [disk-loader](https://github.com/jiangwenyuan/nuster#disk-loader)

2.1 load not complete

When a request comes and if match rule:

- memory on disk off ,  nuster will build the key and lookup the in-memory table, since there's no info in the table, nuster forwards the request to backend
- memory on disk sync|on, nuster will build the key and lookup the in-memory table then try to locate the cache file in disk
- memory off disk on, nuster will locate the cache file in disk

2.2 load complete

After load complete, necessary info are loaded into in-memory table, nuster acts same as 1.

> So during the initialisation of the container (or several containers at the same time) it results into a ton of MISSES and sometimes it even gets lost and starts removing the entire cache folder and starts rebuilding it again.

So there's no MISS during the initialisation if the rebuild is not done while cache is in disk.

**Comment by igorescobar** [comment]
Ok, so that's great.

So if I got it right... there is this transition period that during this step where Nuster is populating the in-memory table (during initialization) it will fallback to disk before forwarding the request to the origin.

Thanks a lot for such detailed explanation @jiangwenyuan 

I hope this discussion helps more people in the future.

**Comment by jiangwenyuan** [comment]
@igorescobar 

> there is this transition period that during this step where Nuster is populating the in-memory table (during initialization) it will fallback to disk before forwarding the request to the origin.

Yes, if there's no entry in the table and the load process is not complete, then check the existence of cache file in disk.

> And even with memory off disk on, Nuster still builds this "in-memory table" with the key -> cache file location, right?

Yes, that's right.

**Comment by igorescobar** [comment]
That's great! 
Thanks!

**Comment by igorescobar** [comment]
Hi,

I'm sorry to bother you again with this but the behaviour still not quite the way you described for me:

This is my `nuster.cfg`:
```
global
    nuster cache on data-size 100g dir /nuster/cache
defaults
    retries 1
    option redispatch
    timeout client  62s
    timeout connect 62s
    timeout server  62s
frontend fe
    bind *:3001
    mode http
    default_backend be
backend be
    mode http
    nuster cache on
    nuster rule r0 disk on memory off use-stale on wait on key method.scheme.host.path
    server s1 web:3000
```

With this configuration, the proxy works and the cache also works but with a few issues:

A few issues:

1. If I restart the container, Nuster keeps returning `503 Service Unavailable` until the `backend` isn't available. I was expecting that it could still return the previously cached file and not have to go to the backend at all.
1. Every time I restart `Nuster` it behaves like it lost all the files it cached on disk.

Worth mentioning that I'm running this test locally using docker-compose:
```
version: "3"
services:
  nuster:
    build: ./terraform/shared/nuster
    volumes:
      - ./terraform/shared/nuster/nuster.conf:/etc/nuster/nuster.conf
      - ./tmp/nuster_cache:/nuster/cache
    ports:
      - 3001:3001
```

Is something wrong with my config?

**Comment by igorescobar** [comment]
I just noticed that everytime I restart the container Nuster is removing the entire content of the cache folder. Any idea why?

First request:
```
nuster_1  | <134>May 19 18:16:41 nuster[7]: {"host":"9fa2367a7924","ident":"haproxy","pid":7,"time":"19/May/2020:18:16:39 +0000","haproxy":{"conn":{"act":1,"fe":1,"be":0,"srv":0},"queue":{"backend":0,"srv":0},"time":{"tq":0,"tw":0,"tc":1,"tr":1819,"tt":1820},"termination_state":"----","retries":0,"network":{"client_ip":"172.18.0.1","client_port":49062,"frontend_ip":"172.18.0.3","frontend_port":3001},"ssl":{"version":"-","ciphers":"-"},"request":{"method":"GET","uri":"\/mylocalhost\/face-04.jpg","protocol":"HTTP/1.1","header":{"host":"-","xforwardfor":"-","referer":"-"}},"name":{"backend":"be","frontend":"fe","server":"s1"},"response":{"status_code":200,"header":{"xrequestid":"-"}},"bytes":{"uploaded":1515,"read":13391}}}
web_1     |   nodemon:watch filterAndRestart on [ 'tmp/nuster_cache/0/0d/0d1cb8d963e6214a62e14d9d081d85926a631310' ]
```

Second request (with cache hit):
```
nuster_1  | <134>May 19 18:17:59 nuster[7]: {"host":"9fa2367a7924","ident":"haproxy","pid":7,"time":"19/May/2020:18:17:59 +0000","haproxy":{"conn":{"act":1,"fe":1,"be":0,"srv":0},"queue":{"backend":0,"srv":0},"time":{"tq":0,"tw":0,"tc":0,"tr":2,"tt":2},"termination_state":"----","retries":0,"network":{"client_ip":"172.18.0.1","client_port":49074,"frontend_ip":"172.18.0.3","frontend_port":3001},"ssl":{"version":"-","ciphers":"-"},"request":{"method":"GET","uri":"\/mylocalhost\/face-04.jpg","protocol":"HTTP/1.1","header":{"host":"-","xforwardfor":"-","referer":"-"}},"name":{"backend":"be","frontend":"fe","server":"<NUSTER.CACHE.ENGINE>"},"response":{"status_code":200,"header":{"xrequestid":"-"}},"bytes":{"uploaded":1515,"read":13391}}}
```

After the restart, the directory gets cleaned up and I get a cache miss.

**Comment by igorescobar** [comment]
I just deployed it to ECS + EFS just to make it wasn't some issue with my local docker setup and the same issue happened. I restarted the container and the EFS volume also got cleaned up.

**Comment by jiangwenyuan** [comment]
@igorescobar Turns out it's a bug introduced when adding stale support, that disk cache are deleted even not expired as long as the `use-stale on` is set. Since the `ttl` is 0 in your config, you can simply remove `use-stale on` as a workaround. I will push the bugfix soon.

Sorry for the inconvenience.

**Comment by igorescobar** [comment]
No worries @jiangwenyuan.
If TTL=0 it means that Nuster will honour my cache http headers, right?

**Comment by jiangwenyuan** [comment]
No, ttl=0 means that the cache does not expire.

Currently nuster does not use cache related headers like `cache-control` .

**Comment by igorescobar** [comment]
Hum,
I was expecting it would work similar to nginx in that case. So it means I need to duplicate my cache definitions on the API onde (needed for the CDN and user browser) and on the Nuster config.

**Comment by jiangwenyuan** [comment]
I understand, my original intention to create nuster was to cache uncacheable request/response for short time in memory, like those( `Cache-Control: no-cache `). So there's no need to support that.

But I plan to add support for that.

**Comment by jiangwenyuan** [comment]
@igorescobar Fixed disk cache delete when use-stale on.

**Comment by igorescobar** [comment]
Hey @jiangwenyuan thanks for the quick fix.

Here it is what's happening now:
1) The good part: the directory is now there and no clean up happens.
2) During the boot of the container, I still have the errors `503`  errors while the backend is not booted.
3) After the backend is online the first request is always a MISS regardless if the file is there already cached or not.

I was expecting that Nuster would  always use the cached file if it available before trying to proxy the request to the backend.

**Comment by igorescobar** [comment]
Ignore my previous comment @jiangwenyuan. I guess it was my mistake.

All good!

Thanks!

Two things that would be nice to have (similar to Nginx):
1) It would be nice if Nuster to add a TTL to files individually based on the Cache-control headers. Nginx respects the header by default but you can turn it off and overwrite the cache definitions.
2) It would be nice to be able to define a "inactive" timer to the cached files. So you can cache the files for what's defined on the headers but if they become inactive for x days you can clean it up.

Cheers!

**Comment by jiangwenyuan** [comment]
@igorescobar Thanks a lot!

> It would be nice if Nuster to add a TTL to files individually based on the Cache-control headers. Nginx respects the header by default but you can turn it off and overwrite the cache definitions.

Yeah, I'm going to add this

> It would be nice to be able to define a "inactive" timer to the cached files. So you can cache the files for what's defined on the headers but if they become inactive for x days you can clean it up.

Can you be more specific? what's the difference with `tt`l and `use-stale`?

**Comment by igorescobar** [comment]
@jiangwenyuan 
> Can you be more specific? What's the difference with ttl and use-stale?
Sure! For example, I can cache a file for 6 months of they are still active/being requested. What the `inactive` does is basically: ttl + inactive would is: Ok, I will cache your file until `ttl` is valid BUT if nobody request your file for more then I will clean it up. It's mostly to save storage and not waste money on dead files. On our use case we want to cache the files for as long as we can, could be months, but it the file is "dead" why would I be storing it.

This is how Nginx currently supports it:
```
proxy_cache_path /path/to/cache levels=1:2 keys_zone=my_cache:10m max_size=10g 
                 inactive=60m use_temp_path=off;
```

**Comment by jiangwenyuan** [comment]
@igorescobar Thanks, say the ttl is 10m, the cache get deleted after 00:10:00, with `inactive 1m`, the cache can be deleted at 00:05:00 if there's no access between 00:04:00~00:05:00 ? Am I understanding correctly?

**Comment by igorescobar** [comment]
@jiangwenyuan It goes like this... "I will cache this file for `ttl` months, but if the file isn't requested for `inactive` days" the disk-cleaner can just delete the file from the because it met the `inactive` criteria.

**Comment by jiangwenyuan** [comment]
@igorescobar I see, thanks, I will implement this.

I'm closing this issue since it's solved, feel free to open issues when you encounter any problems:)

**Comment by igorescobar** [comment]
Thanks a lot @jiangwenyuan !
I can submit other tickets marked "enhancement", it might be better organized for you to track.

**Comment by jiangwenyuan** [comment]
@igorescobar Cool, please:)

**Comment by igorescobar** [comment]
Created!
https://github.com/jiangwenyuan/nuster/issues/78 
https://github.com/jiangwenyuan/nuster/issues/77

Cheers!
